purge_data.py

```diff
@@ -0,0 +1,252 @@
+    isql, dsql, columns_dict = "", "", {}
+
+    if task_runs.empty:
+        return isql, dsql, columns_dict
+
+    for i, task_run in task_runs.iterrows():
+        task_run_data = pd.DataFrame(task_run).transpose()
+        task_run_columns, task_run_col_placeholders, task_run_col_dict = generate_sql_cols_vals(f"tr_{i}",task_run_data)
+        if not(task_run_columns or task_run_col_placeholders or task_run_col_dict):
+            continue
+
+        isql += f"INSERT INTO task_run_archived({task_run_columns}) VALUES({task_run_col_placeholders});"
+        columns_dict.update(task_run_col_dict)
+    dsql = f"DELETE FROM task_run WHERE project_id = %(project_id)s AND task_id = %(task_id)s;"
+    return isql, dsql, columns_dict
+
+
+
+def purge_task_data(task_id, project_id):
+    # make copy of task into respective archived table
+    # delete task data from results, task_runs, task table
+    logger.info(f"Purging task data. project: {project_id}, task: {task_id}")
+    with AccessDatabase() as db:
+        task_data = pd.read_sql_query(f"SELECT * FROM task WHERE id={task_id}", db.conn)
+        task_run_data = pd.read_sql_query(f"SELECT * FROM task_run WHERE task_id={task_id}", db.conn)
+        result_data = pd.read_sql_query(f"SELECT * FROM result WHERE task_id={task_id}", db.conn)
+
+    if task_data.empty:
+        logger.info("Missing data for task id", task_id)
+        logger.info("task_data", task_data)
+        logger.info("task_run_data", task_run_data)
+        logger.info("result_data", result_data)
+        return
+
+    insert_sql, delete_sql = "", ""
+    # archive task data
+    task_columns, task_col_placeholders, task_col_dict = generate_sql_cols_vals("t", task_data)
+    if task_columns and task_col_placeholders and task_col_dict:
+        insert_sql += f"INSERT INTO task_archived({task_columns}) VALUES({task_col_placeholders});"
+        delete_sql += f"DELETE FROM task WHERE project_id = %(project_id)s AND id = %(task_id)s;"
+
+    # archive task_run data
+    import pdb; pdb.set_trace()
+    insert_task_run, delete_task_run, task_run_col_dict = generate_task_run_sql(task_run_data)
+    insert_sql += insert_task_run
+    delete_sql += delete_task_run
+
+    # archive result data
+    result_columns, result_col_placeholders, result_col_dict = generate_sql_cols_vals("r", result_data)
+    if result_columns and result_col_placeholders and result_col_dict:
+        insert_sql += f"INSERT INTO result_archived({result_columns}) VALUES({result_col_placeholders});"
+        delete_sql += f"DELETE FROM result WHERE project_id = %(project_id)s AND task_id = %(task_id)s;"
+
+    # combine task, task_run, result, col-val dict
+    columns_values = {}
+    for d in [task_col_dict, task_run_col_dict, result_col_dict]:
+        columns_values.update(d)
+
+    # add task_id, project_id for delete queries
+    columns_values["task_id"] = task_id
+    columns_values["project_id"] = project_id
+    try:
+        sql = f"""
```

Comment:
Looks like both insert_sql and delete_sql contains multiple sql statement, how long does the insertion and deletion takes? I am wondering if db will kill the process when it exceeds certain timeout value, and the roll back process will be time-consuming too.

Reply:
this script purge and archive tasks per task and not all of the tasks at once, hence not expected to have impact. such script also expected to be run during off business hours.